google/cloud/bigtable/internal/mutate_rows_limiter.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_MUTATE_ROWS_LIMITER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_MUTATE_ROWS_LIMITER_H
+
+#include "google/cloud/bigtable/internal/rate_limiter.h"
+#include "google/cloud/internal/clock.h"
+#include "google/cloud/version.h"
+#include <google/bigtable/v2/bigtable.pb.h>
+#include <chrono>
+#include <functional>
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/// A Bigtable-specific wrapper over the more generic `RateLimiter`
+class MutateRowsLimiter {
+ public:
+  virtual void acquire() = 0;
+  virtual void Update(
+      google::bigtable::v2::MutateRowsResponse const& response) = 0;
```

Comment:
Should this whole thing be divorced from `MutateRows()` by taking a `google::bigtable::v2::RateLimitInfo` instead, just making it `bigtable_internal::RateLimiter`?  `RateLimitInfo` might be used in others calls in the future.  (I guess the mutate-rows `Update()` call would become conditional.)

Reply:
> Should this whole thing be divorced from `MutateRows()`... RateLimitInfo might be used in others calls in the future.

This is a good question. `RateLimitInfo` is defined inside `MutateRowsResponse` though: https://github.com/googleapis/googleapis/blob/92a6b0fc959905f68e567dd60f766167772594bd/google/bigtable/v2/bigtable.proto#L513

The Bigtable team mentioned that the `ReadRows` API *could* be worthy of throttling. But I don't think they plan to add that. This feature is really aimed at heavy write jobs.

Moreover, the flags to enable things are called `mutate_rows_rate_limit[0-9]*`

https://github.com/googleapis/googleapis/blob/92a6b0fc959905f68e567dd60f766167772594bd/google/bigtable/v2/feature_flags.proto#L40-L48

> (I guess the mutate-rows Update() call would become conditional.)

Yeah. This was my main motivation, although it's not a big deal.

Reply:
> `RateLimitInfo` is defined inside `MutateRowsResponse` though

That's not what I see, hence the suggestion.

> Moreover, the flags to enable things are called `mutate_rows_rate_limit[0-9]*`

I don't think that's indicative of anything other than they are the flags for `MutateRows()`.  If `RateLimitInfo` was used elsewhere, presumably that might get its own flags.

>> (I guess the mutate-rows `Update()` call would become conditional.)
> 
> Yeah. This was my main motivation, although it's not a big deal.

I agree that it's not a big deal, and therefore not a good main motivation.  That is, it seems a shame to specialize the interface for that.

Reply:
> That's not what I see, hence the suggestion.

D'oh. I can't read.

Still this class is `internal`. We can easily change it later.

Reply:
> We can easily change it later.

True enough, although I think we suffer a little from we-can-change-it-later-itis.

Reply:
I don't disagree... but observation:

It will take as much effort to change it now as it will to change it later... and we might not need to change it later.

Reply:
We run into trouble when we do this a bunch, and then we **need** to change N of them at the same time. 🫨